Blog post for release 40.0.0 (#6)
apache/datafusion-site: _posts/2024-07-23-datafusion-40.0.0.md

        match &expr.op {
           BinaryOperator::Arrow => { /* plan -> operator */ }
           BinaryOperator::LongArrow => { /* plan ->> operator */ }
           ...
        }
    }
}
```

Thanks to [@samuelcolvin], [@jayzhan211] and [@dharanad] for helping make this
feature happen.

[datafusion-functions-json]: https://github.com/datafusion-contrib/datafusion-functions-json
[`ExprPlanner`]: https://docs.rs/datafusion/latest/datafusion/logical_expr/planner/trait.ExprPlanner.html

## Pluggable Support for `CREATE FUNCTION` 

DataFusion's new [`FunctionFactory`] API let's users provide a handler for
`CREATE FUNCTION` in SQL. This feature lets you build systems that support
defining functions in SQL such as

```sql
-- SQL based functions
CREATE FUNCTION my_func(DOUBLE, DOUBLE) RETURNS DOUBLE
    RETURN $1 + $3
;

-- ML Models
CREATE FUNCTION iris(FLOAT[]) RETURNS FLOAT[] LANGUAGE TORCH AS 'models:/iris@champion';

-- WebAssembly
CREATE FUNCTION func(FLOAT[]) RETURNS FLOAT[] LANGUAGE WASM AS 'func.wasm'
```

Huge thanks to [@milenkovicm] for this feature

BTW it would be great if [someone made a demo] showing how easy this is to do 🎣. 

> [someone made a demo]: https://github.com/apache/datafusion/issues/9326 

Reviewer: Thanks @alamb !
One question, would it make sense to link example https://github.com/apache/datafusion/blob/main/datafusion-examples/examples/function_factory.rs ?

Author: Oh, sweet, yes it would! I think I forgot about that one. Added in https://github.com/apache/datafusion-site/pull/6/commits/79e8f16ec9023328571d87250239a6061572171e

what the community is thinking about in the next 6 months. We are hoping to make
more regular blog posts -- if you are interested in helping write them, please
reach out!

[DataFusion 34.0.0]: https://datafusion.apache.org/blog/2024/01/19/datafusion-34.0.0/
[DataFusion 40.0.0]: https://crates.io/crates/datafusion/40.0.0

[Apache DataFusion] is an extensible query engine, written in [Rust], that
uses [Apache Arrow] as its in-memory format. DataFusion is used by developers to
create new, fast data centric systems such as databases, dataframe libraries,
machine learning and streaming applications. While [DataFusion’s primary design
goal] is to accelerate the creation of other data centric systems, it has a
reasonable experience directly out of the box as a [dataframe library] and
[command line SQL tool].

[DataFusion’s primary design goal]: https://datafusion.apache.org/user-guide/introduction.html#project-goals
[dataframe library]: https://datafusion.apache.org/python/
[command line SQL tool]: https://datafusion.apache.org/user-guide/cli/

[apache datafusion]: https://datafusion.apache.org/
[apache arrow]: https://arrow.apache.org
[rust]: https://www.rust-lang.org/

> DataFusion's core thesis is that as a community together, we can build much more

Reviewer 2: DataFusion's core thesis is that as a community, together we can build much more

[JSON]: https://github.com/datafusion-contrib/datafusion-functions-json
[ORC]: https://github.com/datafusion-contrib/datafusion-orc

<!--
$ git log --pretty=oneline 34.0.0..40.0.0 . | wc -l
     1453 (up from 1009)

$ git shortlog -sn 34.0.0..40.0.0 . | wc -l
      182 (up from 124)

https://crates.io/crates/datafusion/34.0.0
DataFusion 34 released Dec 17, 2023

https://crates.io/crates/datafusion/40.0.0
DataFusion 34 released July 12, 2024

Issues created in this time: 321 open, 781 closed (up from 214 open, 437 closed)
https://github.com/apache/arrow-datafusion/issues?q=is%3Aissue+created%3A2023-12-17..2024-07-12

Issues closed: 911 (up from 517)
https://github.com/apache/arrow-datafusion/issues?q=is%3Aissue+closed%3A2023-12-17..2024-07-12

PRs merged in this time 1490 (up from 908)
https://github.com/apache/arrow-datafusion/pulls?q=is%3Apr+merged%3A2023-12-17..2024-07-12

-->

In addition, DataFusion has been appearing publicly more and more, both online and offline. Here are some highlights:

1. [Apache Arrow DataFusion: A Fast, Embeddable, Modular Analytic Query Engine], was presented in [SIGMOD '24], one of the major database conferences

> 2. As part of the trend to define "the POSIX of databases" in ["What Goes Around Comes Around... And Around...] from Andy Pavlo and Mike Stonebraker

Reviewer 3: ```suggestion
2. As part of the trend to define "the POSIX of databases" in ["What Goes Around Comes Around... And Around..."] from Andy Pavlo and Mike Stonebraker
```
Missing end quote maybe?

We continue to improve the documentation to make it easier to get started using DataFusion with
the [Library Users Guide], [API documentation], and [Examples].

Some notable new examples include:
* [sql_analysis.rs] to analyse SQL queries with DataFusion structures (thanks [@LorrensP-2158466])
* [function_factory.rs] to create custom functions via SQL (thanks [@milenkovicm])
* [plan_to_sql.rs] to generate SQL from DataFusion Expr and LogicalPlan (thanks [@edmondop])
* [parquet_index.rs] and [advanced_parquet_index.rs] for parquet indexing, described more below (thanks [@alamb])

[Library Users Guide]: https://datafusion.apache.org/library-user-guide/index.html
[API documentation]: https://docs.rs/datafusion/latest/datafusion/index.html
[Examples]: https://github.com/apache/datafusion/tree/main/datafusion-examples
[sql_analysis.rs]: https://github.com/apache/datafusion/blob/main/datafusion-examples/examples/sql_analysis.rs
[plan_to_sql.rs]: https://github.com/apache/datafusion/blob/main/datafusion-examples/examples/plan_to_sql.rs

# New Features ✨

There are too many new features in the last 6 months to list them all, but here
are some highlights:

## SQL 

> * Support for`UNNEST` (thanks [@duongcongtoai], [@JasonLi-cn] and [@jayzhan211]) 

Reviewer 3: ```suggestion
* Support for `UNNEST` (thanks [@duongcongtoai], [@JasonLi-cn] and [@jayzhan211]) 
```

  .await?;
// Programmatically add new filters `id > 1 and tinyint_col < double_col`
let df = df.filter(col("id").gt(lit(1)).and(col("tinyint_col").lt(col("double_col"))))?
// Convert the new logical plan back to SQL
let sql = plan_to_sql(df.logical_plan())?.to_string();
assert_eq!(sql, 
           "SELECT alltypes_plain.int_col, alltypes_plain.double_col, CAST(alltypes_plain.date_string_col AS VARCHAR) \
           FROM alltypes_plain WHERE ((alltypes_plain.id > 1) AND (alltypes_plain.tinyint_col < alltypes_plain.double_col))")
);
```

See the [Plan to SQL example] or the APIs [expr_to_sql] and [plan_to_sql] for more details.

[expr_to_sql]: https://docs.rs/datafusion/latest/datafusion/sql/unparser/fn.expr_to_sql.html
[plan_to_sql]: https://docs.rs/datafusion/latest/datafusion/sql/unparser/fn.plan_to_sql.html
[Plan to SQL example]: https://github.com/apache/datafusion/blob/main/datafusion-examples/examples/plan_to_sql.rs

## Low Level APIs for Fast Parquet Access (indexing)

With their rising prevalence, supporting efficient access to Parquet files
stored remotely on object storage is important. Part of doing this efficiently

> is to minimize the number of object stpre requests made by caching metadata and

Reviewer 3: ```suggestion
is to minimize the number of object store requests made by caching metadata and
```

skipping over parts of the file that are not needed (e.g. via an index).

DataFusion's Parquet reader has long internally supported advanced predicate
pushdown by reading the parquet metadata from the file footer and pruning based
on row group level statistics as well as data page level statistics. DataFusion
now also supports users supplying their own low level pruning information via
the [`ParquetAccessPlan`] API.

This API can be used along with index information to selectively skip decoding

> parts of the file. This feature has been used by SpiceAI to add [efficient

Reviewer 4: ```suggestion
parts of the file. This feature has been used by Spice AI to add [efficient
```

Reviewer 4: 🙏

           BinaryOperator::LongArrow => { /* plan ->> operator */ }
           ...
        }
    }
}
```

Thanks to [@samuelcolvin], [@jayzhan211] and [@dharanad] for helping make this
feature happen.

[datafusion-functions-json]: https://github.com/datafusion-contrib/datafusion-functions-json
[`ExprPlanner`]: https://docs.rs/datafusion/latest/datafusion/logical_expr/planner/trait.ExprPlanner.html

## Pluggable Support for `CREATE FUNCTION` 

DataFusion's new [`FunctionFactory`] API let's users provide a handler for
`CREATE FUNCTION` in SQL. This feature lets you build systems that support
defining functions in SQL such as

```sql
-- SQL based functions
CREATE FUNCTION my_func(DOUBLE, DOUBLE) RETURNS DOUBLE
    RETURN $1 + $3
;

-- ML Models
CREATE FUNCTION iris(FLOAT[]) RETURNS FLOAT[] LANGUAGE TORCH AS 'models:/iris@champion';

-- WebAssembly
CREATE FUNCTION func(FLOAT[]) RETURNS FLOAT[] LANGUAGE WASM AS 'func.wasm'
```

Huge thanks to [@milenkovicm] for this feature. There is an example of how to
make macro like functions in the [function_factory.rs] example. It would be
great if [someone made a demo] showing how easy WASM udfs is to do 🎣.

> [fucntion_factory.rs]: https://github.com/apache/datafusion/blob/main/datafusion-examples/examples/function_factory.rs

Reviewer 3: ```suggestion
[function_factory.rs]: https://github.com/apache/datafusion/blob/main/datafusion-examples/examples/function_factory.rs
```
